cpp/log/cert.cc

```diff
@@ -43,6 +43,99 @@ static int X509_get_signature_nid(const X509* x) {
 #endif
 
 
+namespace {
+#if defined(OPENSSL_IS_BORINGSSL)
+// BoringSSL doesn't have DSA hooked up so to accept these certs we have
+// to do the work ourselves. Can be called with a non DSA issuer key but will
+// always return false in that case.
+//
+// cert is the certificate that is to be checked
+// issuer_key is the public key of the certificate issuer (should be DSA)
+// If dsa_key_with_params is not null it's a key that can be used to copy
+// the DSA parameters if they are absent from the issuer key.
+// Returns true if the DSA signature was correctly verified or false if
+// it did not or there was any error along the way.
+StatusOr<bool> check_dsa_signature(const X509* cert, EVP_PKEY* issuer_key,
+                                   const EVP_PKEY* dsa_key_with_params) {
+  // Before we start rummaging about through pointers ensure everything we need
+  // is present
+  if (!cert || !issuer_key || !cert->cert_info || !cert->cert_info->signature ||
+      !cert->cert_info->signature->algorithm) {
+    return ::util::Status(Code::FAILED_PRECONDITION,
+                          "cert is null or missing required parameters");
+  }
+
+  if (issuer_key->type != EVP_PKEY_DSA) {
+    return ::util::Status(Code::FAILED_PRECONDITION,
+                          "cert does not have a DSA public key");
+  }
+
+  // The nid in the cert_info struct seems to be zero so we look it up
+  const int alg_nid = OBJ_obj2nid(cert->cert_info->signature->algorithm);
+
+  int digest_nid;
+  // We need the nid of the digest so we can create an EVP_MD for it later.
+  // Should succeed as we know we have a DSA key
+  if (!OBJ_find_sigid_algs(alg_nid, &digest_nid, nullptr)) {
+    return ::util::Status(Code::INTERNAL, "lookup sigid for algorithm failed");
+  }
+
+  // Get the DER encoded certificate info from the cert
+  unsigned char* der_buf(nullptr);
+  const int der_length = i2d_X509_CINF(cert->cert_info, &der_buf);
+  if (der_length < 0) {
+    // What does this return value mean? Let's assume it means the cert
+    // is bad until proven otherwise. Assume nothing to free.
+    LOG(WARNING) << "Failed to serialize the CINF component";
+    LOG_OPENSSL_ERRORS(WARNING);
+    return ::util::Status(Code::INVALID_ARGUMENT,
+                          "failed to serialize cert info");
+  }
+
+  // If the key is missing parameters try to copy them from the supplied
+  // parent key. TODO: This is not currently enabled as we have no examples
```

Comment:
As I mentioned in the other thread I'd prefer endorsing sane behaviour, and not tolerating everything just because it's in the RFC. IIRC even OpenSSL doesn't let you generate such certs so I doubt anything will.